planner/core/logical_plan_builder.go

```diff
@@ -2671,16 +2675,124 @@ func (b *PlanBuilder) buildProjectionForWindow(p LogicalPlan, expr *ast.WindowFu
 	return proj, propertyItems, newArgList, nil
 }
 
+func (b *PlanBuilder) buildFrameBound(spec *ast.WindowSpec, orderByItems []property.Item, boundClause *ast.FrameBound) (*FrameBound, error) {
+	frameType := spec.Frame.Type
+	bound := &FrameBound{Type: boundClause.Type, UnBounded: boundClause.UnBounded}
+	if bound.UnBounded || boundClause.Type == ast.CurrentRow {
+		return bound, nil
+	}
+
+	if frameType == ast.Rows {
+		// Rows type does not support interval range.
+		if boundClause.Unit != nil {
+			return nil, ErrWindowRowsIntervalUse.GenWithStackByArgs(spec.Name)
+		}
+		numRows, isNull, isExpectedType := getUintFromNode(b.ctx, boundClause.Expr)
+		if isNull || !isExpectedType {
+			return nil, ErrWindowFrameIllegal.GenWithStackByArgs(spec.Name)
+		}
+		bound.Num = numRows
+		return bound, nil
+	}
+
+	if len(orderByItems) != 1 {
+		return nil, ErrWindowRangeFrameOrderType.GenWithStackByArgs(spec.Name)
+	}
+	col := orderByItems[0].Col
+	isNumeric, isTemporal := types.IsTypeNumeric(col.RetType.Tp), types.IsTypeTemporal(col.RetType.Tp)
+	if !isNumeric && !isTemporal {
+		return nil, ErrWindowRangeFrameOrderType.GenWithStackByArgs(spec.Name)
+	}
+	if boundClause.Unit != nil {
+		// Interval bounds only support order by temporal types.
+		if isNumeric {
+			return nil, ErrWindowRangeFrameNumericType.GenWithStackByArgs(spec.Name)
+		}
+
+		// TODO: We also need to raise error for non-deterministic expressions, like rand().
+		val, err := evalAstExpr(b.ctx, boundClause.Expr)
+		if err != nil {
+			return nil, ErrWindowRangeBoundNotConstant.GenWithStackByArgs(spec.Name)
+		}
+		expr := expression.Constant{Value: val, RetType: boundClause.Expr.GetType()}
+		uVal, isNull, err := expr.EvalInt(b.ctx, chunk.Row{})
+		if uVal < 0 || isNull || err != nil {
+			return nil, ErrWindowFrameIllegal.GenWithStackByArgs(spec.Name)
+		}
+		// It can be guaranteed by the parser.
+		unitVal := boundClause.Unit.(*driver.ValueExpr)
+		unit := expression.Constant{Value: unitVal.Datum, RetType: unitVal.GetType()}
+
+		funcName := ast.DateAdd
+		if bound.Type == ast.Preceding {
+			funcName = ast.DateSub
+		}
+		bound.DateCalcFunc, err = expression.NewFunctionBase(b.ctx, funcName, col.RetType, col, &expr, &unit)
```

Comment:
Why we use `NewFunctionBase`? How will it be used?

Reply:
For windows like `order by i_date range interval "2:30" MINUTE_SECOND preceding`, we need to calculate the lower and upper bound by sub/add the interval to the `i_date`, so we need this function.

Reply:
both `NewFunctionBase` and `NewFunction` are okay for this case?

Reply:
Yes, I think so.

Reply:
Would the `bound` become a expression and need `constant fold`?

Reply:
No, it can only be a constant or expressions that can evaluate to constant at the optimization stage. As you can see in previous lines, the `expr` is already evaluated to a constant and `unit` will always be a valid string constant.